coffee/expander.py

```diff
@@ -0,0 +1,244 @@
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import itertools
+
+from base import *
+from utils import *
+
+
+class Cache(object):
+    """A cache for expanded expressions."""
+
+    def __init__(self):
+        self._map = {}
+        self._hits = defaultdict(int)
+
+    def make_key(self, exp, grp):
+        return (str(exp), str(grp))
+
+    def retrieve(self, key):
+        exp = self._map.get(key)
+        if exp:
+            self._hits[key] += 1
+        return exp
+
+    def invalidate(self, exp):
+        for i, j in self._map.items():
+            if str(j) == str(exp):
+                self._map.pop(i)
+                if self._hits[i] > 0:
+                    return True
+        return False
+
+    def add(self, key, exp):
+        self._map[key] = exp
+
+
+class Expander(object):
+
+    """Expand the products in an expression according to a set of rules. For a
+    comprehensive list of possible rules, refer to the documentation of the
+    corresponding wrapper function ``expand`` in ``ExpressionRewriter``."""
+
+    # Constants used by the expand method to charaterize sub-expressions:
+    GROUP = 0  # Expression /will/ not trigger expansion
+    EXPAND = 1  # Expression /could/ be expanded
+
+    # How many times the expander was invoked
+    _handled = 0
+    # Temporary variables template
+    _expanded_sym = "%(loop_dep)s_EXP_%(expr_id)d_%(i)d"
+
+    def __init__(self, stmt, expr_info=None, decls=None, hoisted=None, expr_graph=None):
+        self.stmt = stmt
+        self.expr_info = expr_info
+        self.decls = decls
+        self.hoisted = hoisted
+        self.expr_graph = expr_graph
+
+        self.cache = Cache()
+        self.local_decls = {}
+
+        # Increment counters for unique variable names
+        self.expr_id = Expander._handled
+        Expander._handled += 1
+
+    def _hoist(self, expansion, info):
+        """Try to aggregate an expanded expression E with a previously hoisted
+        expression H. If there are no dependencies, H is expanded with E, so
+        no new symbols need be introduced. Otherwise (e.g., the H temporary
+        appears in multiple places), create a new symbol."""
+        exp, grp = expansion.left, expansion.right
+
+        # First, check if any of the symbols in /exp/ have been hoisted
+        try:
+            retval = FindInstances.default_retval()
+            exp = [s for s in FindInstances(Symbol).visit(exp, ret=retval)[Symbol]
+                   if s.symbol in self.hoisted and self.should_expand(s)][0]
+        except:
```

Comment:
This is very loose, because `except:` will catch expression thrown from anywhere inside this expression. Can you make it more specific?